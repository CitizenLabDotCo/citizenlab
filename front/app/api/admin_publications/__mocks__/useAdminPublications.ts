import { IAdminPublicationData } from '../types';

export const adminPublicationsData: IAdminPublicationData[] = [
  {
    id: 'b3b91b93-e207-4cb0-b49b-737d4b25792e',
    type: 'admin_publication',
    attributes: {
      ordering: 0,
      publication_status: 'published',
      depth: 0,
      publication_title_multiloc: {
        en: 'Quas excepturi ea et.',
        'fr-BE': 'Et illo et aperiam.',
        'nl-BE': 'Praesentium ut itaque ipsam.',
      },
      publication_description_multiloc: {
        en: '\u003cp\u003eAdipisci quia ducimus. Rem et ipsa. Ipsa vero quae.\u003c/p\u003e\u003cp\u003eVoluptas nemo sit. Tempora quia rerum. Ratione dolore quae.\u003c/p\u003e\u003cp\u003eNecessitatibus dolores quod. Magnam ullam natus. Culpa delectus rerum.\u003c/p\u003e',
        'fr-BE':
          '\u003cp\u003eLaudantium et et. Voluptates expedita ipsam. Omnis aut veritatis.\u003c/p\u003e\u003cp\u003eNon cupiditate quia. Illo laudantium tenetur. Neque aperiam aliquid.\u003c/p\u003e\u003cp\u003eAsperiores velit saepe. Veritatis quibusdam est. Fuga laudantium consequatur.\u003c/p\u003e',
        'nl-BE':
          '\u003cp\u003eDeserunt dolor quia. Quae ipsum omnis. Odit autem sint.\u003c/p\u003e\u003cp\u003eEt eos amet. Officia temporibus molestiae. Delectus nihil laboriosam.\u003c/p\u003e\u003cp\u003eMaiores eaque magni. Amet molestias asperiores. Incidunt magni consequatur.\u003c/p\u003e',
      },
      publication_description_preview_multiloc: {
        en: 'Et omnis recusandae minima.',
        'fr-BE': 'Blanditiis et repellendus cupiditate.',
        'nl-BE': 'Libero est voluptatibus tenetur.',
      },
      publication_slug: 'quas-excepturi-ea-et',
      visible_children_count: 1,
    },
    relationships: {
      publication: {
        data: { id: '99d51572-250c-48f4-8950-ea454c9476c1', type: 'folder' },
      },
      parent: { data: null },
      children: {
        data: [
          {
            id: '6012d402-3a56-47a8-8ce6-7cd1c69fa29d',
            type: 'admin_publication',
          },
        ],
      },
    },
  },
  {
    id: '8c8da82d-e4ea-4af4-8c7b-d221e86668cc',
    type: 'admin_publication',
    attributes: {
      ordering: 1,
      publication_status: 'archived',
      depth: 0,
      publication_title_multiloc: {
        en: 'Rem nulla consequatur vero.',
        'fr-BE': 'Ducimus voluptatem tempora expedita.',
        'nl-BE': 'Perferendis modi voluptatem facilis.',
      },
      publication_description_multiloc: {
        en: '\u003cp\u003eAd quam rerum. Fugiat vel dolores. Eligendi occaecati corporis.\u003c/p\u003e\u003cp\u003eDolorem est rem. Sequi aspernatur repudiandae. Tempora reprehenderit aut.\u003c/p\u003e\u003cp\u003eDolor accusantium molestiae. Pariatur aut tempora. Illum sequi illo.\u003c/p\u003e',
        'fr-BE':
          '\u003cp\u003eQuas rem sed. Temporibus corporis et. Fugiat et vero.\u003c/p\u003e\u003cp\u003eEsse aut dolorem. Quis impedit optio. Necessitatibus ipsa ut.\u003c/p\u003e\u003cp\u003eCorporis iure suscipit. Aut autem dolores. Voluptas ea eligendi.\u003c/p\u003e',
        'nl-BE':
          '\u003cp\u003eAliquam amet ut. Ea ut in. Architecto neque minus.\u003c/p\u003e\u003cp\u003eOmnis maxime ullam. Eaque eveniet illum. Animi aut accusantium.\u003c/p\u003e\u003cp\u003eQuasi consectetur dolorem. Ut quam assumenda. Aspernatur pariatur explicabo.\u003c/p\u003e',
      },
      publication_description_preview_multiloc: {
        en: 'Aspernatur ut eius excepturi.',
        'fr-BE': 'Esse incidunt dolore quos.',
        'nl-BE': 'Asperiores possimus labore sequi.',
      },
      publication_slug: 'rem-nulla-consequatur-vero',
      visible_children_count: 1,
    },
    relationships: {
      publication: {
        data: { id: 'c7f4f824-53e1-4b8c-866d-b3412f22d956', type: 'folder' },
      },
      parent: { data: null },
      children: {
        data: [
          {
            id: '8e3a6d01-994e-40d9-9657-9771a100adaf',
            type: 'admin_publication',
          },
        ],
      },
    },
  },
];

export const links = {
  self: 'http://localhost:3000/web_api/v1/admin_publications?depth=0\u0026page%5Bnumber%5D=1\u0026page%5Bsize%5D=1000\u0026publication_statuses%5B%5D=published\u0026publication_statuses%5B%5D=archived\u0026remove_not_allowed_parents=true',
  first:
    'http://localhost:3000/web_api/v1/admin_publications?depth=0\u0026page%5Bnumber%5D=1\u0026page%5Bsize%5D=1000\u0026publication_statuses%5B%5D=published\u0026publication_statuses%5B%5D=archived\u0026remove_not_allowed_parents=true',
  last: 'http://localhost:3000/web_api/v1/admin_publications?depth=0\u0026page%5Bnumber%5D=2\u0026page%5Bsize%5D=1000\u0026publication_statuses%5B%5D=published\u0026publication_statuses%5B%5D=archived\u0026remove_not_allowed_parents=true',
  prev: null,
  next: 'http://localhost:3000/web_api/v1/admin_publications?depth=0\u0026page%5Bnumber%5D=2\u0026page%5Bsize%5D=1000\u0026publication_statuses%5B%5D=published\u0026publication_statuses%5B%5D=archived\u0026remove_not_allowed_parents=true',
};

export default jest.fn(() => {
  return { data: { pages: [{ data: adminPublicationsData, links }] } };
});
