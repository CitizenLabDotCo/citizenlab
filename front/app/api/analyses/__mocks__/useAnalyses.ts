import { IAnalysisData } from '../types';

export const analysesData: IAnalysisData[] = [
  {
    id: '1',
    type: 'analysis',
    attributes: {
      created_at: '2021-06-01T08:00:00.000Z',
      updated_at: '2021-06-01T08:00:00.000Z',
      participation_method: 'survey',
    },
    relationships: {
      project: {
        data: {
          id: '1',
          type: 'project',
        },
      },
      phase: {
        data: {
          id: '1',
          type: 'phase',
        },
      },
      custom_fields: {
        data: [
          {
            id: '1',
            type: 'custom_field',
          },
        ],
      },
    },
  },
];

export default jest.fn(() => {
  return { data: { data: analysesData } };
});
