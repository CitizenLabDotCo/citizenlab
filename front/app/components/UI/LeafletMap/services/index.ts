import { init, addTileLayer, changeView } from './setup';

import { addLayersControl, removeLayersControl } from './controls';
import { addLayers, removeLayers, removeLayer } from './layers';
import {
  addMarkersToMap,
  addMarkerClusterGroup,
  getMarkerIcon,
} from './markers';

export default {
  init,
  addTileLayer,
  changeView,
  addLayersControl,
  removeLayersControl,
  addLayers,
  removeLayers,
  removeLayer,
  addMarkersToMap,
  addMarkerClusterGroup,
  getMarkerIcon,
};
