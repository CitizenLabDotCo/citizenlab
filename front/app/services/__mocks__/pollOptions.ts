export const mockOption = (id, titleEn) => ({
  id,
  attributes: {
    title_multiloc: {
      en: titleEn,
    },
  },
});
