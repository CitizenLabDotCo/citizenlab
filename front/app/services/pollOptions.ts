import { API_PATH } from 'containers/App/constants';
import streams from 'utils/streams';
import { Multiloc } from 'typings';

export function updatePollOption(optionId: string, titleMultiloc: Multiloc) {
  return streams.update(`${API_PATH}/poll_options/${optionId}`, optionId, {
    title_multiloc: titleMultiloc,
  });
}
