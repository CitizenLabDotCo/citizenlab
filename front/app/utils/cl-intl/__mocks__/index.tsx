import React from 'react';
const Intl = jest.requireActual('react-intl');

// Initialise the real provider so that we don't
// need to reimplement any internals
const defaultProps = {
  locale: 'en',
  defaultLocale: 'en',
};
const intlProvider = new Intl.IntlProvider(defaultProps, {});

// The exact same `intl` object the real code receives ;-)
const { intl } = intlProvider.getChildContext();

const formatMessageReplacement = (
  messageDescriptor: ReactIntl.FormattedMessage.MessageDescriptor,
  values?: { [key: string]: string | number | boolean | Date } | undefined
) => {
  return intl.formatMessage(messageDescriptor, {
    tenantName: 'The Test',
    orgName: 'Test Town',
    orgType: 'testing',
    ...(values || {}),
  });
};

const intlReplacement = {
  ...intl,
  formatMessage: formatMessageReplacement,
};

const injectIntlReplacement = (Node) => {
  const renderWrapped = (props) => <Node {...props} intl={intl} />;
  renderWrapped.displayName = Node.displayName || Node.name || 'Component';
  return renderWrapped;
};

Intl.injectIntl = injectIntlReplacement;

// Override components by implementing the real components
//  providing them the context they need in order to function
const {
  IntlProvider,
  FormattedDate,
  FormattedTime,
  FormattedRelative,
  FormattedNumber,
  FormattedPlural,
  FormattedMessage,
  FormattedHTMLMessage,
} = Intl;

Intl.FormattedDate = (props) => (
  <IntlProvider {...defaultProps}>
    <FormattedDate {...props} />
  </IntlProvider>
);
Intl.FormattedTime = (props) => (
  <IntlProvider {...defaultProps}>
    <FormattedTime {...props} />
  </IntlProvider>
);
Intl.FormattedRelative = (props) => (
  <IntlProvider {...defaultProps}>
    <FormattedRelative {...props} />
  </IntlProvider>
);
Intl.FormattedNumber = (props) => (
  <IntlProvider {...defaultProps}>
    <FormattedNumber {...props} />
  </IntlProvider>
);
Intl.FormattedPlural = (props) => (
  <IntlProvider {...defaultProps}>
    <FormattedPlural {...props} />
  </IntlProvider>
);
Intl.FormattedMessage = (props) => (
  <IntlProvider {...defaultProps}>
    <FormattedMessage {...props} />
  </IntlProvider>
);
Intl.FormattedHTMLMessage = (props) => (
  <IntlProvider {...defaultProps}>
    <FormattedHTMLMessage {...props} />
  </IntlProvider>
);

// Set displayName so that snapshots don't use "Uknown" as component name
Intl.FormattedDate.displayName = 'FormattedDate';
Intl.FormattedTime.displayName = 'FormattedTime';
Intl.FormattedRelative.displayName = 'FormattedRelative';
Intl.FormattedNumber.displayName = 'FormattedNumber';
Intl.FormattedPlural.displayName = 'FormattedPlural';
Intl.FormattedMessage.displayName = 'FormattedMessage';
Intl.FormattedHTMLMessage.displayName = 'FormattedHTMLMessage';

// Special hook for tests, real package does not export this
Intl.intl = intlReplacement;

module.exports = Intl;
