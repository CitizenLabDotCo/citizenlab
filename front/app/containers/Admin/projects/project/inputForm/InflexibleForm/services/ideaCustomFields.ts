import { API_PATH } from 'containers/App/constants';
import streams, { IStreamParams } from 'utils/streams';
import { Multiloc } from 'typings';
import { CustomFieldCodes } from 'services/ideaCustomFieldsSchemas';

export type Visibility = 'admins' | 'public';

export interface IIdeaCustomFieldData {
  id: string;
  type: 'custom_field';
  attributes: {
    key: string;
    input_type:
      | 'text'
      | 'number'
      | 'multiline_text'
      | 'html'
      | 'text_multiloc'
      | 'multiline_text_multiloc'
      | 'html_multiloc'
      | 'select'
      | 'multiselect'
      | 'checkbox'
      | 'date'
      | 'files'
      | 'image_files';
    title_multiloc: Multiloc;
    description_multiloc: Multiloc;
    required: boolean;
    ordering: null;
    enabled: boolean;
    code: CustomFieldCodes;
    created_at: null;
    updated_at: null;
    visible_to: Visibility;
  };
}

export interface IIdeaCustomField {
  data: IIdeaCustomFieldData;
}

export interface IIdeaCustomFields {
  data: IIdeaCustomFieldData[];
}

export interface IUpdatedIdeaCustomFieldProperties {
  description_multiloc?: Multiloc;
  enabled?: boolean;
  required?: boolean;
  visible_to?: Visibility;
}

export function ideaCustomFieldsStream(
  projectId: string,
  streamParams: IStreamParams | null = null
) {
  const apiEndpoint = `${API_PATH}/admin/projects/${projectId}/custom_fields`;
  return streams.get<IIdeaCustomFields>({ apiEndpoint, ...streamParams });
}

export async function updateIdeaCustomField(
  projectId: string,
  ideaCustomFieldId: string,
  code: string | undefined,
  object: IUpdatedIdeaCustomFieldProperties
) {
  const apiEndpoint = code
    ? `${API_PATH}/admin/projects/${projectId}/custom_fields/by_code/${code}`
    : `${API_PATH}/admin/projects/${projectId}/custom_fields/update/${ideaCustomFieldId}`;
  const updateObject = { custom_field: object };
  return streams.update<IIdeaCustomField>(
    apiEndpoint,
    ideaCustomFieldId,
    updateObject
  );
}

export async function refetchCustomFields(projectId) {
  const apiEndpoint = `${API_PATH}/admin/projects/${projectId}/custom_fields`;
  streams.fetchAllWith({ apiEndpoint: [apiEndpoint] });
}
