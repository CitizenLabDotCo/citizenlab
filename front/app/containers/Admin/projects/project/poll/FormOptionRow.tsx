// Libraries
import React, { PureComponent } from 'react';
import { adopt } from 'react-adopt';
import { isNilOrError } from 'utils/helperUtils';

// Components
import { Input, LocaleSwitcher } from '@citizenlab/cl2-component-library';
import { TextCell, Row } from 'components/admin/ResourceList';
import Button from 'components/UI/Button';

// Resources
import GetLocale, { GetLocaleChildProps } from 'resources/GetLocale';
import GetAppConfigurationLocales, {
  GetAppConfigurationLocalesChildProps,
} from 'resources/GetAppConfigurationLocales';

// Typings
import { Multiloc, Locale } from 'typings';

// Services
import { addPollOption, updatePollOption } from 'services/pollOptions';

// i18n
import { FormattedMessage } from 'utils/cl-intl';
import messages from './messages';

/*
 * edit mode : titleMultiloc and optionId defined, question Id not used
 * new mode : question Id defined, titleMultiloc and optionId not used
 */
interface InputProps {
  titleMultiloc?: Multiloc;
  mode: 'new' | 'edit';
  questionId?: string;
  closeRow: () => void;
  optionId?: string;
}

interface DataProps {
  locale: GetLocaleChildProps;
  tenantLocales: GetAppConfigurationLocalesChildProps;
}

export interface Props extends DataProps, InputProps {}

export interface State {
  selectedLocale: Locale | null;
  titleMultiloc: Multiloc;
}

export class FormOptionRow extends PureComponent<Props, State> {
  constructor(props: Props) {
    super(props);
    this.state = {
      selectedLocale: props.locale || null,
      titleMultiloc: props.titleMultiloc || {},
    };
  }

  static getDerivedStateFromProps(props: Props, state: State) {
    if (!isNilOrError(props.locale) && !state.selectedLocale) {
      return {
        selectedLocale: props.locale,
      };
    }

    return null;
  }

  componentDidUpdate(prevProps: Props) {
    if (prevProps.optionId !== this.props.optionId) {
      this.setState({ titleMultiloc: this.props.titleMultiloc || {} });
    }
  }

  onSelectedLocaleChange = (selectedLocale: Locale) => {
    this.setState({ selectedLocale });
  };

  onChangeTitle = (value: string, locale: Locale | undefined) => {
    if (locale) {
      this.setState((state) => ({
        titleMultiloc: {
          ...state.titleMultiloc,
          [locale]: value,
        },
      }));
    }
  };

  onSave = () => {
    const { mode, questionId, closeRow, optionId } = this.props;
    const { titleMultiloc } = this.state;

    if (mode === 'new' && questionId) {
      addPollOption(questionId, titleMultiloc).then(() => {
        closeRow();
      });
    }

    if (mode === 'edit' && optionId) {
      updatePollOption(optionId, titleMultiloc).then(() => {
        closeRow();
      });
    }
  };

  render() {
    const { selectedLocale, titleMultiloc } = this.state;
    const { closeRow, tenantLocales } = this.props;

    return (
      <Row className="e2e-form-option-row">
        <TextCell>
          {selectedLocale && (
            <LocaleSwitcher
              onSelectedLocaleChange={this.onSelectedLocaleChange}
              locales={!isNilOrError(tenantLocales) ? tenantLocales : []}
              selectedLocale={selectedLocale}
              values={{ titleMultiloc }}
            />
          )}
        </TextCell>

        <TextCell className="expand">
          {selectedLocale && (
            <Input
              autoFocus
              value={titleMultiloc?.[selectedLocale]}
              locale={selectedLocale}
              type="text"
              onChange={this.onChangeTitle}
            />
          )}
        </TextCell>

        <Button
          className="e2e-form-option-save"
          buttonStyle="secondary"
          onClick={this.onSave}
        >
          <FormattedMessage {...messages.saveOption} />
        </Button>

        <Button
          className="e2e-form-option-cancel"
          buttonStyle="secondary"
          onClick={closeRow}
        >
          <FormattedMessage {...messages.cancelOption} />
        </Button>
      </Row>
    );
  }
}

const Data = adopt<DataProps, InputProps>({
  locale: <GetLocale />,
  tenantLocales: <GetAppConfigurationLocales />,
});

const FormOptionRowWithData = (inputProps: InputProps) => (
  <Data {...inputProps}>
    {(dataProps) => <FormOptionRow {...dataProps} {...inputProps} />}
  </Data>
);

export default FormOptionRowWithData;
