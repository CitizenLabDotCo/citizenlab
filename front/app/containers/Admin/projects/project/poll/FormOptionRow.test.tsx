import React, { Component } from 'react';
import { shallow } from 'enzyme';
import { mockOption } from 'services/__mocks__/pollOptions';

jest.mock('services/pollOptions', () => ({
  addPollOption: jest.fn((_id, _type, title) => {
    return new Promise((resolve) =>
      resolve({ data: mockOption('newOption', title) })
    );
  }),
  updatePollOption: jest.fn(() => {
    return new Promise((resolve) => resolve);
  }),
}));

import * as pollOptionsService from 'services/pollOptions';
const addPollOptionSpy = jest.spyOn(pollOptionsService, 'addPollOption');
const updatePollOptionSpy = jest.spyOn(pollOptionsService, 'updatePollOption');

import { Input, LocaleSwitcher } from '@citizenlab/cl2-component-library';

jest.mock('@citizenlab/cl2-component-library', () => ({
  Input: 'Input',
  LocaleSwitcher: 'LocaleSwitcher',
}));
jest.mock('components/admin/ResourceList', () => ({
  TextCell: 'TextCell',
  Row: 'Row',
}));
jest.mock(
  'components/UI/InputMultilocWithLocaleSwitcher',
  () => 'InputMultilocWithLocaleSwitcher'
);
jest.mock('components/UI/Button', () => 'Button');
jest.mock('utils/cl-intl', () => ({ FormattedMessage: 'FormattedMessage' }));

import { FormOptionRow, Props, State } from './FormOptionRow';

let closeRow = jest.fn();
const getTitleMultiloc = (title: string) => ({ en: title });

describe('<FormOptionRow />', () => {
  beforeEach(() => {
    closeRow = jest.fn();
  });

  it('matches the snapshot', () => {
    const titleMultiloc = getTitleMultiloc('Vanilla');
    const wrapper = shallow<Component<Props, State>>(
      <FormOptionRow
        titleMultiloc={titleMultiloc}
        closeRow={closeRow}
        mode="edit"
        questionId="questionId"
        optionId="optionId"
        locale="en"
        tenantLocales={['en', 'fr-BE']}
      />
    );
    expect(wrapper).toMatchSnapshot();
  });

  describe('boundaries', () => {
    it('reacts to option change', () => {
      const titleMultiloc = getTitleMultiloc('Vanilla');
      const wrapper = shallow<Component<Props, State>>(
        <FormOptionRow
          titleMultiloc={titleMultiloc}
          closeRow={closeRow}
          mode="edit"
          questionId="questionId"
          optionId="optionId"
          locale="en"
          tenantLocales={['en', 'fr-BE']}
        />
      );
      wrapper.setProps({
        titleMultiloc: getTitleMultiloc('Pistachio'),
        optionId: 'anotherOption',
      });
      expect(wrapper.find('Input').prop('value')).toEqual('Pistachio');
    });
  });

  describe('handles language switch for multilingual content', () => {
    it('shows the passed in locale by default', () => {
      const wrapper = shallow<Component<Props, State>>(
        <FormOptionRow
          closeRow={closeRow}
          mode="new"
          questionId="questionId"
          locale="en"
          tenantLocales={['en', 'fr-BE']}
        />
      );
      expect(wrapper.find('Input').prop('locale')).toBe('en');
      expect(wrapper.find('LocaleSwitcher').prop('selectedLocale')).toBe('en');
    });

    it('handles changing field locale', () => {
      const wrapper = shallow<Component<Props, State>>(
        <FormOptionRow
          titleMultiloc={getTitleMultiloc('Vanilla')}
          closeRow={closeRow}
          mode="new"
          questionId="questionId"
          locale="en"
          tenantLocales={['en', 'fr-BE']}
        />
      );
      wrapper.find(LocaleSwitcher).prop('onSelectedLocaleChange')('fr-BE');
      expect(wrapper.find('Input').prop('locale')).toBe('fr-BE');
    });
  });

  describe('handles input of title multiloc', () => {
    describe('for a new option', () => {
      it('passes down initial value', () => {
        const wrapper = shallow<Component<Props, State>>(
          <FormOptionRow
            closeRow={closeRow}
            mode="new"
            questionId="questionId"
            locale="en"
            tenantLocales={['en', 'fr-BE']}
          />
        );
        expect(wrapper.find('Input').prop('value')).toEqual(undefined);
      });

      it('reacts to user input', () => {
        const titleMultiloc = getTitleMultiloc('Vanilla');
        const wrapper = shallow<Component<Props, State>>(
          <FormOptionRow
            titleMultiloc={titleMultiloc}
            closeRow={closeRow}
            mode="new"
            questionId="questionId"
            locale="en"
            tenantLocales={['en', 'fr-BE']}
          />
        );
        wrapper.find(Input).prop('onChange' as any)('Pistachio', 'en');
        wrapper.find(Input).prop('onChange' as any)('Chocolate', 'fr-BE');
        const instance = wrapper.instance();
        expect(instance.state.titleMultiloc['en']).toBe('Pistachio');
        expect(instance.state.titleMultiloc['fr-BE']).toBe('Chocolate');
      });
    });

    describe('when editing an option', () => {
      it('passes down initial value', () => {
        const titleMultiloc = getTitleMultiloc('Vanilla');
        const wrapper = shallow<Component<Props, State>>(
          <FormOptionRow
            titleMultiloc={titleMultiloc}
            closeRow={closeRow}
            mode="edit"
            questionId="questionId"
            optionId="optionId"
            locale="en"
            tenantLocales={['en', 'fr-BE']}
          />
        );
        expect(wrapper.find('Input').prop('value')).toEqual(
          titleMultiloc['en']
        );
      });

      it('reacts to user input', () => {
        const wrapper = shallow<Component<Props, State>>(
          <FormOptionRow
            titleMultiloc={getTitleMultiloc('Vani')}
            closeRow={closeRow}
            mode="edit"
            questionId="questionId"
            optionId="optionId"
            locale="en"
            tenantLocales={['en', 'fr-BE']}
          />
        );
        wrapper.find(Input).prop('onChange' as any)('Vanilla', 'en');
        expect(wrapper.find('Input').prop('value')).toEqual('Vanilla');
      });
    });
  });

  describe('handles saving', () => {
    describe('for a new option', () => {
      it('handles saving the new option', () => {
        const titleMultiloc = getTitleMultiloc('Vani');
        const questionId = 'questionId';
        const wrapper = shallow<Component<Props, State>>(
          <FormOptionRow
            titleMultiloc={getTitleMultiloc('Vani')}
            closeRow={closeRow}
            mode="new"
            questionId={questionId}
            locale="en"
            tenantLocales={['en', 'fr-BE']}
          />
        );
        wrapper.find('.e2e-form-option-save').simulate('click');
        expect(addPollOptionSpy).toHaveBeenCalledTimes(1);
        expect(addPollOptionSpy).toHaveBeenCalledWith(
          questionId,
          titleMultiloc
        );
        // resolve the promise the test the callback
        return pollOptionsService
          .addPollOption(questionId, titleMultiloc)
          .then(() => {
            expect(closeRow).toHaveBeenCalledTimes(1);
          });
      });
    });

    describe('when editing an option', () => {
      it('handles updating the option', () => {
        const wrapper = shallow<Component<Props, State>>(
          <FormOptionRow
            closeRow={closeRow}
            mode="edit"
            questionId="questionId"
            optionId="optionId"
            locale="en"
            tenantLocales={['en', 'fr-BE']}
          />
        );
        wrapper.setState({ titleMultiloc: getTitleMultiloc('Vanilla') });
        wrapper.find('.e2e-form-option-save').simulate('click');
        expect(updatePollOptionSpy).toHaveBeenCalledTimes(1);
        expect(updatePollOptionSpy).toHaveBeenCalledWith('optionId', {
          en: 'Vanilla',
        });
      });
    });
  });

  describe('handles cancelling', () => {
    describe('for a new option', () => {
      it('handles closing the form to cancel', () => {
        const wrapper = shallow<Component<Props, State>>(
          <FormOptionRow
            closeRow={closeRow}
            mode="new"
            questionId="questionId"
            locale="en"
            tenantLocales={['en', 'fr-BE']}
          />
        );
        wrapper.find('.e2e-form-option-cancel').simulate('click');
        expect(closeRow).toHaveBeenCalledTimes(1);
      });
    });

    describe('when editing an option', () => {
      it('handles closing the form to cancel', () => {
        const wrapper = shallow<Component<Props, State>>(
          <FormOptionRow
            titleMultiloc={getTitleMultiloc('Vani')}
            closeRow={closeRow}
            mode="new"
            questionId="questionId"
            optionId="optionId"
            locale="en"
            tenantLocales={['en', 'fr-BE']}
          />
        );
        wrapper.find('.e2e-form-option-cancel').simulate('click');
        expect(closeRow).toHaveBeenCalledTimes(1);
      });
    });
  });
});
