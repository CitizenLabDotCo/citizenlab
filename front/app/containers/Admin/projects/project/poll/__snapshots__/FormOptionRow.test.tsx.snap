// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<FormOptionRow /> matches the snapshot 1`] = `
<Row
  className="e2e-form-option-row"
>
  <TextCell>
    <LocaleSwitcher
      locales={
        Array [
          "en",
          "fr-BE",
        ]
      }
      onSelectedLocaleChange={[Function]}
      selectedLocale="en"
      values={
        Object {
          "titleMultiloc": Object {
            "en": "Vanilla",
          },
        }
      }
    />
  </TextCell>
  <TextCell
    className="expand"
  >
    <Input
      autoFocus={true}
      locale="en"
      onChange={[Function]}
      type="text"
      value="Vanilla"
    />
  </TextCell>
  <Button
    buttonStyle="secondary"
    className="e2e-form-option-save"
    onClick={[Function]}
  >
    <FormattedMessage
      defaultMessage="Save"
      id="app.containers.AdminPage.ProjectEdit.PollTab.saveOption"
    />
  </Button>
  <Button
    buttonStyle="secondary"
    className="e2e-form-option-cancel"
    onClick={[MockFunction]}
  >
    <FormattedMessage
      defaultMessage="Cancel"
      id="app.containers.AdminPage.ProjectEdit.PollTab.cancelOption"
    />
  </Button>
</Row>
`;
